### PR TITLE
support margin loss (arcface, cosface, sphereface) for single GPU and cross GPUs

### PR DESCRIPTION
**Pull request by GuoxiaWang**
### PR types
 New features

### PR changes
APIs

### Describe
margin loss, it is a unified loss, which includes arcface, cosface, sphereface

-m_{3})}}{e^{s(cos(m_{1}\\theta_{y_i}+m_{2})-m_{3})}+\\sum^n_{j=1,j\\neq%20y_i}e^{scos\\theta_{y_i}}})

The details of margin loss could be referred to https://arxiv.org/abs/1801.07698.

Note that the APIs support model parallel and single GPU. And Logits.shape[-1] can be different each rank.

### CODE REVIEW

File: python/paddle/nn/functional/loss.py [hunk 1]
```diff
@@ -1092,6 +1092,260 @@ def ctc_loss(log_probs,
...
+                    If :attr:`reduction` is ``'mean'``, return the average of loss;
+                    If :attr:`reduction` is ``'sum'``, return the sum of loss;
+                    If :attr:`reduction` is ``'none'``, no reduction will be applied.
+
+    Return:
+    	loss (Tensor or Scalar): if reduction==None, shape[N, 1], else shape[1], the cross entropy loss.
+    	softmax (Tensor): softmax probability. The softmax is shard_softmax when using model parallel.
+
+    Examples:
+    .. code-block:: python
```
Comment (wadefelix) [inline python/paddle/nn/functional/loss.py:1143]: 此行后，应有一个空行。

**Review by wadefelix — CHANGES_REQUESTED** [review]
code-block指令后应有一个空行

File: python/paddle/nn/functional/loss.py (hunk 1 above)
Comment (GuoxiaWang) [reply python/paddle/nn/functional/loss.py:1143]: > 此行后，应有一个空行。

好的，谢谢

**Review by sandyhouse — DISMISSED** [review]
LGTM

**Review by sandyhouse — DISMISSED** [review]
LGTM

**Review by TCChenlong — DISMISSED** [review]
LGTM

**Review by sandyhouse — DISMISSED** [review]
LGTM

**Review by XiaoguangHu01 — DISMISSED** [review]
LG API

**Review by sandyhouse — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LG API

**Review by TCChenlong — APPROVED** [review]

**Review by raindrops2sea — APPROVED** [review]

**Review by kolinwei — APPROVED** [review]

**Review by chenwhql — APPROVED** [review]
LGTM for PADDLE_ENFORCE

**Review by zhiqiu — APPROVED** [review]
LGTM for ShareDataWith

**Review by XieYunshen — APPROVED** [review]
LGTM

**Pull request merged**